models/exposure.py

```diff
@@ -220,43 +250,92 @@ class Exposure(Base, FileOnDiskMixin, SpatiallyIndexed):
+        """Initialize the exposure object.
+
+        If the filepath is given (as a keyword argument), it will parse the instrument name
+        from the filename.  The header will be read out from the file.
 
-        Otherwise, give any arguments that are
-        columns of the Exposure table.
+        Parameters
+        ----------
+        All the properties of Exposure (i.e. columns of the exposures table), plus:
+
+        current_file: Path or str
+           The path to the file where the exposure currently is (which
+           may or may not be the same as the filepath it will have in
+           the database).  If you don't specify this, then the file must
+           exist filepath (either the one you pass or the one that is
+           determined automatically if invent_filepath is True).
+
+        invent_filepath: bool
+           Will be ignored if you specify a filepath as an argument.
+           Otherwise, if this is True, call invent_filepath() to create
+           the filepath for this Exposure, based on all the other
+           properties.  If this is False, then you must specify filepath
+           unless the global property Exposure.nofile is True (but you
+           really shouldn't be playing around with that).
 
-        If the filename is given, it will parse
-        the instrument name from the filename.
-        The header will be read out from the file.
         """
-        FileOnDiskMixin.__init__(self, *args, **kwargs)
+        FileOnDiskMixin.__init__(self, **kwargs)
         SeeChangeBase.__init__(self)  # don't pass kwargs as they could contain non-column key-values
 
         self._data = None  # the underlying image data for each section
         self._section_headers = None  # the headers for individual sections, directly from the FITS file
         self._raw_header = None  # the global (exposure level) header, directly from the FITS file
         self.type = 'Sci'  # default, can override using kwargs
 
-        if self.filepath is None and not self.nofile:
-            raise ValueError("Must give a filepath to initialize an Exposure object. ")
+        # manually set all properties (columns or not, but don't
+        # overwrite instance methods) Do this once here, because some of
+        # the values are going to be needed by upcoming function calls.
+        # (See "chicken and egg" comment below).  We will run this exact
+        # code again later so that the keywords can override what's
+        # detected from the header.
+        for key, value in kwargs.items():
+            if hasattr(self, key):
+                if type( getattr( self, key ) ) != types.MethodType:
+                    setattr(self, key, value)
+
+        # a default provenance for exposures
+        if self.provenance is None:
+            codeversion = Provenance.get_code_version()
+            self.provenance = Provenance.create_or_load( code_version=codeversion, process='load_exposure' )
```

Comment:
So, thinking about this: as things stand right now, we are going to need something like a "create or load" method for Provenance that goes to the database.  Reason: if you just create a Provenance with the constructor, and do things, when you get to committing it to the database, a simple session.merge() will not merge it properly, as the new provenance you created won't have the same primary key as the other provenance that has the unique hash in the database.   That's the whole reason I wrote this; if you ask for a provenance that already exists, you need to get the already existing one.  There's no way to do that without either using a session or creating a session to connect to the database.

There is another solution, which would be my preference.  That would be to take the id field out of Base in models/base.py.  Add a HasPrimaryID mixin that has the auto-incrementing id.  Most models would include that, but Provenance would not; for Provenance, the unique hash would *be* the primary key.  Provenance doesn't need an autoincrementing primary key, since the unique hash already serves any purpose a primary key would need.  Then, we could create provenances by just making objects, and when we merge them with the session it would merge correctly.

Reply:
That's an interesting suggestion. I need to think about that.

Reply:
I agree that you need to check the database for Provenance. There's a tool to do that on the data_store object. 
The problem is that this is not a simple thing you can just have happen automatically in the constructor of the parent class. You need to know the upstream provenance and all the parameters.